core/src/main/java/org/pentaho/di/core/logging/LoggingRegistry.java

```diff
@@ -0,0 +1,718 @@
+        }
+      }
+
+      return determineLogChannelFileWriterBuffer( possibleWriters );
+    }
+  }
+
+  /**
+   * Helper Method that determines a LogChannelFileWriterBuffer invoked by getLogChannelFileWriterBuffer and returns 1.
+   * @param possibleWriters  Map to search from.
+   * @return LogChannelFileWriterBuffer, null if could not be determined.
+   */
+  private LogChannelFileWriterBuffer determineLogChannelFileWriterBuffer( ConcurrentHashMap<LogChannelFileWriterBuffer,
+    List<String>> possibleWriters ) {
+
+    // Just one writer so just return it
+    if ( possibleWriters.size() == 1 ) {
+      return possibleWriters.keys().nextElement();
+    } else {
+
+      // Several possibilities, so, lets get the writer among them that is the "lowest in the chain",
+      // meaning, the one that is not a parent of the others
+      Enumeration<LogChannelFileWriterBuffer> possibleWritersIds = possibleWriters.keys();
+      while ( possibleWritersIds.hasMoreElements() ) {
+        LogChannelFileWriterBuffer writer = possibleWritersIds.nextElement();
+
+        for ( Map.Entry<LogChannelFileWriterBuffer, List<String>> entry : possibleWriters.entrySet() ) {
+          if ( entry.getKey().equals( writer ) ) {
+            continue;
+          }
+          if ( !entry.getValue().contains( writer.getLogChannelId() ) ) {
+            return entry.getKey();
+          }
+        }
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns a Set of LogChannelIds associated with a FileWriterBuffer which includes a parent and its children.
+   * @return  A set of LogChannelIds to avoid duplicates.
+   */
+  protected Set<String> getLogChannelFileWriterBufferIds() {
+    Set<String> bufferIds = this.fileWriterBuffers.keySet();
+
+    // Changed to a set as a band-aid for PDI-16658. This stuff really should be done
+    // using a proper LRU cache.
+    Set<String> ids = new HashSet<>();
+    for ( String id : bufferIds ) {
+      ids.addAll( getLogChannelChildren( id ) );
+    }
+
+    ids.addAll( bufferIds );
+    return ids;
+  }
+
+  /**
+   * Removes a LogChannelFileWriterBuffer by LogChannelId.
+   * @param id  LogChannelId of the object to remove.
+   */
+  public void removeLogChannelFileWriterBuffer( String id ) {
+    synchronized ( this.syncObject ) {
+      Set<String> bufferIds = this.fileWriterBuffers.keySet();
+      for ( String bufferId : bufferIds ) {
+        if ( getLogChannelChildren( id ).contains( bufferId ) ) {
+          this.fileWriterBuffers.remove( bufferId );
+        }
+      }
+    }
+  }
+
+  /**
+   * Resets all internal memory objects and counters.
+   */
+  public void reset() {
+    synchronized ( this.syncObject ) {
+      map.clear();
+      childrenMap.clear();
+      fileWriterBuffers.clear();
+      registerPurgeQueue.clear();
+
+      purgeTimerCount = 0;
+      purgedObjectCount = 0;
+      foundCounter = 0;
+
+
+      if ( purgeTimer != null ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It seems unusual to set the value of a variable only when it is already set. You might want to switch the null check of != on line 471 to == check before setting the field value on line 474 or verify if the check is needed at all.